In [1]:
import pandas as pd
import numpy as np
data=pd.read_csv('hs_data_all_fields.csv', index_col=0)
data.drop(labels=['Freshmen_On_Track_CPS_Pct_Year_2','Graduation_4_Year_School_Pct_Year_2','Graduation_4_Year_School_Pct_Year_1','Graduation_5_Year_School_Pct_Year_1','College_Enrollment_CPS_Pct_Year_2','College_Enrollment_School_Pct_Year_1','College_Persistence_CPS_Pct_Year_2','Mobility_Rate_Pct','College_Persistence_School_Pct_Year_1','College_Persistence_CPS_Pct_Year_1','College_Enrollment_CPS_Pct_Year_1','Graduation_5_Year_CPS_Pct_Year_1','Graduation_5_Year_CPS_Pct_Year_2','Graduation_4_Year_CPS_Pct_Year_1','Graduation_4_Year_CPS_Pct_Year_2','Freshmen_On_Track_School_Pct_Year_1','Freshmen_On_Track_CPS_Pct_Year_1','Unnamed: 0.1','School_Hours','Legacy_Unit_ID','Finance_ID','Short_Name','Summary','Primary_Category','Is_High_School','Is_Middle_School','Is_Elementary_School','Is_Pre_School','Administrator_Title','Administrator','Phone','Secondary_Contact_Title','Secondary_Contact','Address','City','State','Zip','Fax','Freshman_Start_End_Time','After_School_Hours','Earliest_Drop_Off_Time','Significantly_Modified','Blue_Ribbon_Award_Year','Spot_Light_Award_Year','Improvement_Award_Year','Excellence_Award_Year','Hard_Of_Hearing','Mean_ACT','College_Enrollment_Rate_Mean','School_Year','Progress_Report_Year','Graduation_Rate_Mean','Visual_Impairments','School_Latitude','Classroom_Languages','CPS_School_Profile','Website','Facebook','Twitter','Youtube','Pinterest','Attendance_Boundaries','Grades_Offered_All','Grades_Offered','Student_Count_Native_American','Student_Count_Other_Ethnicity','Student_Count_Asian_Pacific_Islander','Student_Count_Multi','Student_Count_Hawaiian_Pacific_Islander','Student_Count_Ethnicity_Not_Available','Statistics_Description','Demographic_Description','PreK_School_Day','Kindergarten_School_Day','PreSchool_Inclusive','Preschool_Instructional','School_Longitude','School_Longitude','Rating_Statement','Classification_Description','Third_Contact_Title','Third_Contact_Name','Fourth_Contact_Title','Fourth_Contact_Name','Fifth_Contact_Title','Fifth_Contact_Name','Sixth_Contact_Title','Sixth_Contact_Name','Seventh_Contact_Title','Seventh_Contact_Name','Location','Excelerate_Award_Gold_Year','Student_Growth_Description','Growth_Reading_Grades_Tested_Pct_ES','Growth_Reading_Grades_Tested_Label_ES','Growth_Math_Grades_Tested_Pct_ES','Growth_Math_Grades_Tested_Label_ES','Student_Attainment_Description','Attainment_Reading_Pct_ES','Attainment_Reading_Lbl_ES','Attainment_Math_Pct_ES','Attainment_Math_Lbl_ES','Culture_Climate_Description','School_Survey_Student_Response_Rate_Avg_Pct','School_Survey_Teacher_Response_Rate_Avg_Pct','Healthy_School_Certification','Healthy_School_Certification_Description','School_Survey_Student_Response_Rate_Pct','School_Survey_Teacher_Response_Rate_Pct','Creative_School_Certification','Creative_School_Certification_Description','NWEA_Reading_Growth_Grade_3_Pct','NWEA_Reading_Growth_Grade_3_Lbl','NWEA_Reading_Growth_Grade_4_Pct','NWEA_Reading_Growth_Grade_4_Lbl','NWEA_Reading_Growth_Grade_5_Pct','NWEA_Reading_Growth_Grade_5_Lbl','NWEA_Reading_Growth_Grade_6_Pct','NWEA_Reading_Growth_Grade_6_Lbl','NWEA_Reading_Growth_Grade_7_Pct','NWEA_Reading_Growth_Grade_7_Lbl','NWEA_Reading_Growth_Grade_8_Pct','NWEA_Reading_Growth_Grade_8_Lbl','NWEA_Math_Growth_Grade_3_Pct','NWEA_Math_Growth_Grade_3_Lbl','NWEA_Math_Growth_Grade_4_Pct','NWEA_Math_Growth_Grade_4_Lbl','NWEA_Math_Growth_Grade_5_Pct','NWEA_Math_Growth_Grade_5_Lbl','NWEA_Math_Growth_Grade_6_Pct','NWEA_Math_Growth_Grade_6_Lbl','NWEA_Math_Growth_Grade_7_Pct','NWEA_Math_Growth_Grade_7_Lbl','NWEA_Math_Growth_Grade_8_Pct','NWEA_Math_Growth_Grade_8_Lbl','NWEA_Reading_Attainment_Grade_2_Pct','NWEA_Reading_Attainment_Grade_2_Lbl','NWEA_Reading_Attainment_Grade_3_Pct','NWEA_Reading_Attainment_Grade_3_Lbl','NWEA_Reading_Attainment_Grade_4_Pct','NWEA_Reading_Attainment_Grade_4_Lbl','NWEA_Reading_Attainment_Grade_5_Pct','NWEA_Reading_Attainment_Grade_5_Lbl','NWEA_Reading_Attainment_Grade_6_Pct','NWEA_Reading_Attainment_Grade_6_Lbl','NWEA_Reading_Attainment_Grade_7_Pct','NWEA_Reading_Attainment_Grade_7_Lbl','NWEA_Reading_Attainment_Grade_8_Pct','NWEA_Reading_Attainment_Grade_8_Lbl','NWEA_Math_Attainment_Grade_2_Pct','NWEA_Math_Attainment_Grade_2_Lbl','NWEA_Math_Attainment_Grade_3_Pct','NWEA_Math_Attainment_Grade_3_Lbl','NWEA_Math_Attainment_Grade_4_Pct','NWEA_Math_Attainment_Grade_4_Lbl','NWEA_Math_Attainment_Grade_5_Pct','NWEA_Math_Attainment_Grade_5_Lbl','NWEA_Math_Attainment_Grade_6_Pct','NWEA_Math_Attainment_Grade_6_Lbl','NWEA_Math_Attainment_Grade_7_Pct','NWEA_Math_Attainment_Grade_7_Lbl','NWEA_Math_Attainment_Grade_8_Pct','NWEA_Math_Attainment_Grade_8_Lbl','Suspensions_Per_100_Students_Avg_Pct','Misconducts_To_Suspensions_Avg_Pct','Average_Length_Suspension_Year_1_Pct','Average_Length_Suspension_Year_2_Pct','Average_Length_Suspension_Avg_Pct','Behavior_Discipline_Year_1','Behavior_Discipline_Year_2','Student_Attendance_Avg_Pct','Teacher_Attendance_Avg_Pct','One_Year_Dropout_Rate_Avg_Pct','Other_Metrics_Year_1','Other_Metrics_Year_2','Growth_ACT_Grade_11_Lbl','Attainment_ACT_Grade_11_Lbl','Progress_Toward_Graduation_Year_1','Progress_Toward_Graduation_Year_2','State_School_Report_Card_URL','Chronic_Truancy_Pct','Empty_Progress_Report_Message','School_Survey_Rating_Description','Supportive_School_Award_Desc','Parent_Survey_Results_Year','School_Latitude.1','School_Longitude.1','Transportation_Bus','Transportation_El','Transportation_Metra'], axis=1,inplace=True)
data.set_index('School_ID', inplace=True)
name=data['Long_Name']
data.drop(labels='Long_Name', axis=1, inplace=True)
data=data[-data['Graduation_Rate_School'].isna()]
data.describe(include='all').T

count unique  \
School_Type                                    121     10   
Student_Count_Total                            121    NaN   
Student_Count_Low_Income                       121    NaN   
Student_Count_Special_Ed                       121    NaN   
Student_Count_English_Learners                 121    NaN   
Student_Count_Black                            121    NaN   
Student_Count_Hispanic                         121    NaN   
Student_Count_White                            121    NaN   
Student_Count_Asian                            121    NaN   
ADA_Accessible                                 121      3   
Dress_Code                                     121      2   
Bilingual_Services                             115      2   
Refugee_Services                               113      2   
Title_1_Eligible                               121      2   
Average_ACT_School                             117    NaN   
College_Enrollment_Rate_School                 119    NaN   
Graduation_Rate_School                         121    NaN   
Overall_Rating                                 121      6   
Rating_Status                                  121      4   
Student_Growth_Rating                          121      6   
Student_Attainment_Rating                      121      6   
Culture_Climate_Rating                         121      6   
School_Survey_Involved_Families                121      6   
School_Survey_Supportive_Environment           121      5   
School_Survey_Ambitious_Instruction            121      6   
School_Survey_Effective_Leaders                121      6   
School_Survey_Collaborative_Teachers           121      5   
School_Survey_Safety                           121      6   
Suspensions_Per_100_Students_Year_1_Pct         88    NaN   
Suspensions_Per_100_Students_Year_2_Pct         88    NaN   
Misconducts_To_Suspensions_Year_1_Pct           88    NaN   
Misconducts_To_Suspensions_Year_2_Pct           88    NaN   
School_Survey_School_Community                 121      4   
School_Survey_Parent_Teacher_Partnership       121      5   
School_Survey_Quality_Of_Facilities            121      5   
Student_Attendance_Year_1_Pct                  121    NaN   
Student_Attendance_Year_2_Pct                  121    NaN   
Teacher_Attendance_Year_1_Pct                   85    NaN   
Teacher_Attendance_Year_2_Pct                   85    NaN   
One_Year_Dropout_Rate_Year_1_Pct               119    NaN   
One_Year_Dropout_Rate_Year_2_Pct               119    NaN   
Growth_ACT_Grade_11_Pct                        114    NaN   
Attainment_ACT_Grade_11_Pct                    117    NaN   
Freshmen_On_Track_School_Pct_Year_2            117    NaN   
Graduation_5_Year_School_Pct_Year_2            121    NaN   
College_Enrollment_School_Pct_Year_2           119    NaN   
College_Persistence_School_Pct_Year_2          115    NaN   
Supportive_School_Award                        121      4   
lib_cnt                                        121    NaN   
Neighborhood Name                              121     57   
PRED_NON_ENG_LANG                              121      4   
PRED_NON_ENG_PERCENT                           121    NaN   
PERCENT OF HOUSING CROWDED                     121    NaN   
PERCENT HOUSEHOLDS BELOW POVERTY               121    NaN   
PERCENT AGED 16+ UNEMPLOYED                    121    NaN   
PERCENT AGED 25+ WITHOUT HIGH SCHOOL DIPLOMA   121    NaN   
PERCENT AGED UNDER 18 OR OVER 64               121    NaN   
PER CAPITA INCOME                              121    NaN   
HARDSHIP INDEX                                 121    NaN   

                                                               top freq  \
School_Type                                           Neighborhood   44   
Student_Count_Total                                            NaN  NaN   
Student_Count_Low_Income                                       NaN  NaN   
Student_Count_Special_Ed                                       NaN  NaN   
Student_C

In the above step we are deleting irrelevant column names, subsetting the data to school that have graduation rate value filled out, separating names of schools for future use if needed, changing index to school id. 

Next step is to create change values from fields that have year 1 and year 2 and drop original variables

In [2]:
data['Suspensions_Per_100_Students_Change']=data['Suspensions_Per_100_Students_Year_1_Pct']-data['Suspensions_Per_100_Students_Year_2_Pct']
data['Misconducts_To_Suspensions_Change']=data['Misconducts_To_Suspensions_Year_1_Pct']-data['Misconducts_To_Suspensions_Year_2_Pct']
data['Student_Attendance_Change']=data['Student_Attendance_Year_1_Pct']-data['Student_Attendance_Year_2_Pct']
data['Teacher_Attendance_Change']=data['Teacher_Attendance_Year_1_Pct']-data['Teacher_Attendance_Year_2_Pct']
data['One_Year_Dropout_Rate_Change']=data['One_Year_Dropout_Rate_Year_1_Pct']-data['One_Year_Dropout_Rate_Year_2_Pct']
data.drop(labels=['College_Enrollment_School_Pct_Year_2','Graduation_5_Year_School_Pct_Year_2','Suspensions_Per_100_Students_Year_1_Pct','Suspensions_Per_100_Students_Year_2_Pct','Misconducts_To_Suspensions_Year_1_Pct','Misconducts_To_Suspensions_Year_2_Pct','Student_Attendance_Year_1_Pct','Student_Attendance_Year_2_Pct','Teacher_Attendance_Year_1_Pct','Teacher_Attendance_Year_2_Pct','One_Year_Dropout_Rate_Year_1_Pct','One_Year_Dropout_Rate_Year_2_Pct'],axis=1,inplace=True)

Following step fills in null values:

average for missing values of ACT Score, college enrollment rate, and other percent values related to school or college performance

zeroes for rates of change from previous year; we are assuming that schools that had missing values for suspension, misconduct, teacher attendance and one year drop out rates, had no changces from 2015 to 2016

U for unknown for categorical values that otherwise have a Y/N indicator

In [3]:
data['Average_ACT_School'].fillna(data['Average_ACT_School'].mean(),inplace=True)
data['College_Enrollment_Rate_School'].fillna(data['College_Enrollment_Rate_School'].mean(),inplace=True)
data['Growth_ACT_Grade_11_Pct'].fillna(data['Growth_ACT_Grade_11_Pct'].mean(),inplace=True)
data['Attainment_ACT_Grade_11_Pct'].fillna(data['Attainment_ACT_Grade_11_Pct'].mean(),inplace=True)
data['Freshmen_On_Track_School_Pct_Year_2'].fillna(data['Freshmen_On_Track_School_Pct_Year_2'].mean(),inplace=True)
data['College_Persistence_School_Pct_Year_2'].fillna(data['College_Persistence_School_Pct_Year_2'].mean(),inplace=True)

data['Suspensions_Per_100_Students_Change'].fillna(0,inplace=True)
data['Misconducts_To_Suspensions_Change'].fillna(0,inplace=True)
data['Teacher_Attendance_Change'].fillna(0,inplace=True)
data['One_Year_Dropout_Rate_Change'].fillna(0,inplace=True)

data['Bilingual_Services'].fillna('U',inplace=True)
data['Refugee_Services'].fillna('U',inplace=True)


Some values in the data set had counts of students in the school that were part of a group (low income, special ed, different ethnic groups, etc). We calculated rates for those groups using provided counts and student total count for the school. Then, dropped the original count columns.

In [4]:
data['Student_Low_Income_Pct']=data['Student_Count_Low_Income']/data['Student_Count_Total']
data['Student_Special_Ed_Pct']=data['Student_Count_Special_Ed']/data['Student_Count_Total']
data['Student_English_Learners_Pct']=data['Student_Count_English_Learners']/data['Student_Count_Total']
data['Student_Black_Pct']=data['Student_Count_Black']/data['Student_Count_Total']
data['Student_Hispanic_Pct']=data['Student_Count_Hispanic']/data['Student_Count_Total']
data['Student_White_Pct']=data['Student_Count_White']/data['Student_Count_Total']
data['Student_Asian_Pct']=data['Student_Count_Asian']/data['Student_Count_Total']

data.drop(labels=['Student_Count_Low_Income','Student_Count_Special_Ed','Student_Count_English_Learners','Student_Count_Black','Student_Count_Hispanic','Student_Count_White','Student_Count_Asian'], axis=1,inplace=True)

In the following step we are creating a new dataset with dummies for categorical variables

In [5]:
data_dummies=pd.get_dummies(data)

Next step is to create another dataset with all variables scaled from 0 to 1. This data set will be used for regression modeling.

In [6]:
from sklearn.preprocessing import MinMaxScaler
min_max=MinMaxScaler()
min_max.fit(data_dummies)
data_scaled=pd.DataFrame(min_max.transform(data_dummies),index=data_dummies.index.values,columns=data_dummies.columns.values)

pd.set_option('display.max_columns', 100)
data_scaled.describe(include='all').T

count      mean       std  min  \
Student_Count_Total                           121.0  0.165612  0.155159  0.0   
Average_ACT_School                            121.0  0.264698  0.187867  0.0   
College_Enrollment_Rate_School                121.0  0.626671  0.202823  0.0   
Graduation_Rate_School                        121.0  0.745314  0.198001  0.0   
Growth_ACT_Grade_11_Pct                       121.0  0.363839  0.206636  0.0   
Attainment_ACT_Grade_11_Pct                   121.0  0.277600  0.255489  0.0   
Freshmen_On_Track_School_Pct_Year_2           121.0  0.745342  0.191599  0.0   
College_Persistence_School_Pct_Year_2         121.0  0.534013  0.221170  0.0   
lib_cnt                                       121.0  0.390684  0.197660  0.0   
PRED_NON_ENG_PERCENT                          121.0  0.255280  0.324058  0.0   
PERCENT OF HOUSING CROWDED                    121.0  0.344282  0.277461  0.0   
PERCENT HOUSEHOLDS BELOW POVERTY              121.0  0.388523  0.199614  0.0   
PERCENT AGED 16+ UNEMPLOYED                   121.0  0.361173  0.230247  0.0   
PERCENT AGED 25+ WITHOUT HIGH SCHOOL DIPLOMA  121.0  0.378996  0.259615  0.0   
PERCENT AGED UNDER 18 OR OVER 64              121.0  0.543954  0.207344  0.0   
PER CAPITA INCOME                             121.0  0.212112  0.190222  0.0   
HARDSHIP INDEX                                121.0  0.531737  0.314494  0.0   
Suspensions_Per_100_Students_Change           121.0  0.497658  0.094905  0.0   
Misconducts_To_Suspensions_Change             121.0  0.460626  0.161228  0.0   
Student_Attendance_Change                     121.0  0.130348  0.094966  0.0   
Teacher_Attendance_Change                     121.0  0.527326  0.352256  0.0   
One_Year_Dropout_Rate_Change                  121.0  0.445522  0.127153  0.0   
Student_Low_Income_Pct                        121.0  0.808827  0.207230  0.0   
Student_Special_Ed_Pct                        121.0  0.169600  0.139432  0.0   
Student_English_Learners_Pct                  121.0  0.204971  0.245138  0.0   
Student_Black_Pct                             121.0  0.519915  0.388664  0.0   
Student_Hispanic_Pct                          121.0  0.387639  0.347367  0.0   
Student_White_Pct                             121.0  0.120364  0.212846  0.0   
Student_Asian_Pct                             121.0  0.083004  0.170391  0.0   
School_Type_Career academy                    121.0  0.033058  0.179531  0.0   
...                                             ...       ...       ...  ...   
Neighborhood Name_MOUNT GREENWOOD             121.0  0.008264  0.090909  0.0   
Neighborhood Name_NEAR NORTH SIDE             121.0  0.008264  0.090909  0.0   
Neighborhood Name_NEAR SOUTH SIDE             121.0  0.008264  0.090909  0.0   
Neighborhood Name_NEAR WEST SIDE              121.0  0.074380  0.263480  0.0   
Neighborhood Name_NEW CITY                    121.0  0.016529  0.128028  0.0   
Neighborhood Name_NORTH CENTER                121.0  0.016529  0.128028  0.0   
Neighborhood Name_NORTH LAWNDALE              121.0  0.024793  0.156141  0.0   
Neighborhood Name_NORTH PARK                  121.0  0.033058  0.179531  0.0   
Neighborhood Name_NORWOOD PARK                121.0  0.008264  0.090909  0.0   
Neighborhood Name_PORTAGE PARK                121.0  0.016529  0.128028  0.0   
Neighborhood Name_PULLMAN                     121.0  0.008264  0.090909  0.0   
Neighborhood Name_RIVERDALE                   121.0  0.008264  0.090909  0.0   
Neighborhood Name_ROGERS PARK                 121.0  0.016529  0.128028  0.0   
Neighborhood Name_ROSELAND                    121.0  0.024793  0.156141  0.0   
Neighborhood Name_SOUTH CHICAGO               121.0  0.016529  0.128028  0.0   
Neighborhood Name_SOUTH LAWNDALE              121.0  0.057851  0.234433  0.0   
Neighborhood Name_SOUTH SHORE                 121.0  0.008264  0.090909  0.0   
Neighborhood Name_UPTOWN                      121.0  0.008264  0.090909  0.0   
Neighborhood Name_WASHINGTON HEIGHTS          

Checking description of all variables I can make sure that all variable have a count of 121, with min of 0 and max of 1.

Last three steps export three datasets for use in analysis and modeling.

In [7]:
data_dummies.to_csv('processed_data.csv')

In [8]:
data.to_csv('processed_nodummies.csv')

In [9]:
data_scaled.to_csv('processed_scaled.csv')